src/opf/dcp.jl

```diff
@@ -1,263 +1,179 @@
-    length(ref[:ref_buses]) == 1 || error("Data has $(length(ref[:ref_buses])) slack buses (expected 1).")
-    i0 = first(keys(ref[:ref_buses]))
+    # TODO: remove when all formulations are done
+    data = OPFData(network)
 
     # Grab some data
-    N = length(ref[:bus])
-    G = length(ref[:gen])
-    E = length(data["branch"])
-    L = length(ref[:load])
-    bus_loads = [
-        [ref[:load][l] for l in ref[:bus_loads][i]]
-        for i in 1:N
-    ]
-    bus_shunts = [
-        [ref[:shunt][s] for s in ref[:bus_shunts][i]]
-        for i in 1:N
-    ]
+    N, E, G = data.N, data.E, data.G
+    gs = data.gs
+    pd = data.pd
+    i0 = data.ref_bus
+    bus_arcs_fr, bus_arcs_to = data.bus_arcs_fr, data.bus_arcs_to
+    bus_gens = data.bus_gens
+    pgmin, pgmax = data.pgmin, data.pgmax
+    c0, c1, c2 = data.c0, data.c1, data.c2
+    gen_status = data.gen_status
+    bus_fr, bus_to, b = data.bus_fr, data.bus_to, data.b
+    dvamin, dvamax, smax = data.dvamin, data.dvamax, data.smax
+    branch_status = data.branch_status
 
     model = JuMP.GenericModel{T}(optimizer)
-    model.ext[:opf_model] = DCOPF  # for internal checks
+    model.ext[:opf_model] = DCOPF
 
     #
     #   I. Variables
     #
+    
+    # nodal voltage
+    @variable(model, va[1:N])
 
-    # bus voltage angle
-    JuMP.@variable(model, va[1:N])
-
-    # generator active dispatch
-    JuMP.@variable(model, ref[:gen][g]["pmin"] <= pg[g in 1:G] <= ref[:gen][g]["pmax"])
+    # Active and reactive dispatch
+    @variable(model, pg[g in 1:G])
 
-    # branch flows
-    JuMP.@variable(model, -ref[:branch][l]["rate_a"] <= pf[(l,i,j) in ref[:arcs_from]] <= ref[:branch][l]["rate_a"])
+    # Branch flows
+    @variable(model, pf[e in 1:E])
 
-    # https://github.com/lanl-ansi/PowerModels.jl/blob/d9c00d228a4e875ac3426425ad6c8f8338309efa/src/form/dcp.jl#L255-L258
-    # pf_expr = Dict{Any, Any}( ((l,i,j), pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from] )
-    # pf_expr = merge(pf_expr, Dict( ((l,j,i), -1.0*pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from] ))
-    pf_expr = [((l,i,j), pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from]]
-    pf_expr = vcat(pf_expr, [((l,j,i), -1.0*pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from]])
-
-    model[:pf] = pf = JuMP.Containers.DenseAxisArray(collect(getfield.(pf_expr, 2)), collect(getfield.(pf_expr, 1)))
-
-    #
+    # 
     #   II. Constraints
     #
 
+    # Generation bounds (both zero if generator is off)
+    set_lower_bound.(pg, gen_status .* pgmin)
+    set_upper_bound.(pg, gen_status .* pgmax)
+
+    # Flow bounds (both zero if branch is off)
+    set_lower_bound.(pf, branch_status .* -smax)
+    set_upper_bound.(pf, branch_status .* smax)
+
     # Slack bus
-    JuMP.@constraint(model, slack_bus, va[i0] == 0.0)
+    @constraint(model, slack_bus, va[i0] == 0.0)
 
     # Nodal power balance
-    JuMP.@constraint(model,
+    @expression(model, pt[e in 1:E], -pf[e])
+    @constraint(model,
         kirchhoff[i in 1:N],
```

Comment:
For consistency with the other formulations
```suggestion
        kcl_p[i in 1:N],
```